### PR TITLE
enh(Dashboards): added health calculation in Status grid widget for Impact BAs

### PR DESCRIPTION
**Pull request by dmyios**
## Description

Enhanced health calculation in Status grid widget for impact BAs to display the current level of a BA and color it depending on the current level.

image

**Fixes** # MON-152749

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [ ] 24.04.x
- [ ] 24.10.x
- [x] master

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/useBATooltipContent.ts [hunk 1]
```diff
@@ -58,7 +63,9 @@ const useBATooltipContent = (
 
   const ProblematicKPIsCount = indicatorsWithProblems?.length || 0;
 
-  const health = Math.floor(((total - ProblematicKPIsCount) * 100) / total);
+  const health = equals(calculationMethod, CalculationMethodType.Impact) 
+    ? businessActivity?.currentLevel || 0
+    : Math.floor(((total - ProblematicKPIsCount) * 100) / total);
```
Comment (Med1233) [inline centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/useBATooltipContent.ts:68]: I suppose that the health is only used for impact CM

**Comment by Med1233** [comment]
Can you please compare the values of thresholds with the values in the configuration page?  IMO 125% is not a logic value for the warning level.

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/useBATooltipContent.ts (hunk 1 above)
Comment (dmyios) [reply centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/useBATooltipContent.ts:68]: Yes, the API will return the current level value only for BAs of type Impact

**Comment by dmyios** [comment]
> the values of thresholds with the values in the configuration p

The thing is you can set the warning level to 125% in configuration page. IMO the limitation should be done when configuring a BA (not allowing the user to set thresholds more than 100%). The values on the screenshot were set for the demonstration sake.

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/useBATooltipContent.ts (hunk 1 above)
Comment (dmyios) [reply centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/useBATooltipContent.ts]: Tru. I've corrected it

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx [hunk 2]
```diff
@@ -76,6 +76,21 @@ const BATooltipContent = ({ data }: Props): JSX.Element | null => {
...
+      }
+      if (health <= warningLevel) {
+        return SeverityCode.Medium;
+      }
+
+      return SeverityCode.OK;
+    }
+
+    return SeverityCode.High;
+  };
```
Comment (Med1233) [inline centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx:92]:
```suggestion
const getHealthSeverityCode = (): SeverityCode => {
  if (isNil(criticalLevel) || isNil(warningLevel) || health <= criticalLevel) {
    return SeverityCode.High;
  }

  if (health <= warningLevel) {
    return SeverityCode.Medium;
  }

  return SeverityCode.OK;
};
```

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx [hunk 3]
```diff
@@ -121,7 +136,7 @@ const BATooltipContent = ({ data }: Props): JSX.Element | null => {
                   <Typography
                     sx={{
                       color: getColor({
-                        severityCode: isImpact ? 5 : 1,
+                        severityCode: calculateSeverityCode(),
```
Comment (Med1233) [inline centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx:139]:
```suggestion
                        severityCode: isImpact ? getHealthSeverityCode(): 1,
```

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx (hunk 2 above)
Comment (Med1233) [reply centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx:92]: prevent nested conditions

**Review by Med1233 — APPROVED** [review]

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx [hunk 4]
```diff
@@ -76,6 +76,18 @@ const BATooltipContent = ({ data }: Props): JSX.Element | null => {
...
+  if (isNil(criticalLevel) || isNil(warningLevel) || health <= criticalLevel) {
+    return SeverityCode.High;
+  }
+
+  if (health <= warningLevel) {
+    return SeverityCode.Medium;
+  }
+
+  return SeverityCode.OK;
+};
```
Comment (Thebarda) [inline centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx:89]: Change it to a value `healthSeverityCode` and wrap it around a useMemo

Comment (dmyios) [reply centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx]: Done

File: centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx [hunk 5]
```diff
@@ -76,17 +77,22 @@ const BATooltipContent = ({ data }: Props): JSX.Element | null => {
...
 
-  return SeverityCode.OK;
-};
+    return SeverityCode.OK;
+  };
+
+  const healthSeverityCode = useMemo(
+    getHealthSeverityCode,
+    [criticalLevel, warningLevel, health]
+  );
```
Comment (Thebarda) [inline centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/Widgets/centreon-widget-statusgrid/src/StatusGridStandard/Tooltip/BATooltipContent.tsx:95]:
```
const healthSeverityCode = useMemo(
    () => {
      if (isNil(criticalLevel) || isNil(warningLevel) || health <= criticalLevel) {
      return SeverityCode.High;
    }
if (health <= warningLevel) {
      return SeverityCode.Medium;
    }
return SeverityCode.OK;
    },
    [criticalLevel, warningLevel, health]
  );
```

**Review by quborg — DISMISSED** [review]

**Review by Thebarda — APPROVED** [review]

**Review by Noha-ElAbrouki — APPROVED** [review]

**Pull request merged**